cmake: Provide a clear error on version check fail

Aleksbgbg opened the pull request:
### Description
when git describe fails to get a git tag for the OBS Version, a non-fatal message is printed, and the generator is left to continue, usually ending up with a more cryptic error message down the line, such as in #7705.

Instead, dump the git output together with a short description of what actually happened, and exit fatally so the problematic line of code is clear. An added advantage is that the git output is printed in red instead of (say) white on colour-enabled terminals.

### Motivation and Context
Gives developers a more clear error for debugging issues like #7705.

### How Has This Been Tested?
Tested with CMake 3.25 on Debian 12.

Examples:

#### Before:
```
$ cmake -S . -B YOUR_BUILD_DIRECTORY -G Ninja  -DCEF_ROOT_DIR="../obs-deps/cef_binary_5060_linux_x86_64"       -DENABLE_PIPEWIRE=OFF        -DENABLE_AJA=0         -DENABLE_WEBRTC=1        -DQT_VERSION=6                                               
-- OBS:  Application Version:  - Build Number: 16                                                                                                                                                                                                                                         
CMake Warning at CMakeLists.txt:46 (project):                                                                                                                                                                                                                                             
  VERSION keyword not followed by a value or was followed by a value that                                                                                                                                                                                                                 
  expanded to nothing.                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                          
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so                                                                                                                                                                                                                                   
-- Found FFmpeg: /usr/lib/x86_64-linux-gnu/libavcodec.so (found version "#define LIBAVCODEC_VERSION_MINOR  37;#define LIBAVCODEC_VERSION_MICRO 100.37.100") found components: avcodec avdevice avutil avformat                                                                            
-- OBS:  ENABLED    obslua                                                                                                                                                                                                                                                                
-- OBS:  obs-scripting -> Luajit found.                                                                                                                                                                                                                                                   
-- OBS:  ENABLED    obspython                                                                                                                                                                                                                                                             
-- OBS:  obs-scripting -> Python 3.11.2 found.
-- XCB[XCB]: Found component XCB                                      
-- Found XCB: /usr/lib/x86_64-linux-gnu/libxcb.so  found components: XCB 
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libOpenGL.so  found components: EGL 
-- Found Wayland: /usr/lib/x86_64-linux-gnu/libwayland-client.so;/usr/lib/x86_64-linux-gnu/libwayland-server.so;/usr/lib/x86_64-linux-gnu/libwayland-egl.so;/usr/lib/x86_64-linux-gnu/libwayland-cursor.so   
-- Found FFmpeg: /usr/lib/x86_64-linux-gnu/libavformat.so (found version "#define LIBAVFORMAT_VERSION_MINOR  27;#define LIBAVFORMAT_VERSION_MICRO 100.27.100") found components: avformat avutil swscale swresample avcodec 
-- OBS:  -> PulseAudio found - audio monitoring enabled
-- Found Wayland: /usr/lib/x86_64-linux-gnu/libwayland-client.so;/usr/lib/x86_64-linux-gnu/libwayland-server.so;/usr/lib/x86_64-linux-gnu/libwayland-egl.so;/usr/lib/x86_64-linux-gnu/libwayland-cursor.so  found components: Client 
CMake Error at /usr/share/cmake-3.25/Modules/WriteBasicConfigVersionFile.cmake:43 (message):
  No VERSION specified for WRITE_BASIC_CONFIG_VERSION_FILE()
Call Stack (most recent call first):
  /usr/share/cmake-3.25/Modules/CMakePackageConfigHelpers.cmake:237 (write_basic_config_version_file)
  cmake/Modules/ObsHelpers.cmake:269 (write_basic_package_version_file)
  cmake/Modules/ObsHelpers_Linux.cmake:13 (_export_target)
  libobs/cmake/legacy.cmake:502 (export_target)
  cmake/Modules/ObsHelpers.cmake:473 (include)
  libobs/CMakeLists.txt:3 (legacy_check)


-- Configuring incomplete, errors occurred!
See also "/src/obs-studio/YOUR_BUILD_DIRECTORY/CMakeFiles/CMakeOutput.log".
```

#### After:
```
$ cmake -S . -B YOUR_BUILD_DIRECTORY -G Ninja  -DCEF_ROOT_DIR="../obs-deps/cef_binary_5060_linux_x86_64"       -DENABLE_PIPEWIRE=OFF        -DENABLE_AJA=0         -DENABLE_WEBRTC=1        -DQT_VERSION=6 
CMake Error at cmake/Modules/VersionConfig.cmake:26 (message):
  Could not fetch OBS version tag from git.

  fatal: detected dubious ownership in repository at '/src/obs-studio'

  To add an exception for this directory, call:

  

        git config --global --add safe.directory /src/obs-studio

Call Stack (most recent call first):
  CMakeLists.txt:38 (include)


-- Configuring incomplete, errors occurred!
See also "/src/obs-studio/YOUR_BUILD_DIRECTORY/CMakeFiles/CMakeOutput.log".
```

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
- Code cleanup (non-breaking change which makes code smaller or more readable)
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX reviewed (changes requested):
I believe that this is the old CMake path, which is currently only really used on Linux. This must also be updated in the new CMake path, which is slated to become the default on all platforms in the future.

Please also apply the updates to this file:
* cmake/common/versionconfig.cmake

Please also use an appropriate commit message (prefix+subject) per our [Commit Guidelines](https://github.com/obsproject/obs-studio/blob/1badf8e11b2f4beb6adb93580e4662ebff52bea1/CONTRIBUTING.rst#commit-guidelines). In this case, use the `cmake: ` prefix.

Aleksbgbg commented:
Thanks for the review, fixed.

I also added a new commit which will fail the version check when tags are unavailable (by removing `--always`), and fallback to the error message introduced by this change. Now when I manually delete my tags I get this error message:
```
λ cmake --preset windows-x64
Preset CMake variables:

  ENABLE_BROWSER:BOOL="TRUE"
  ENABLE_CCACHE:BOOL="FALSE"
  GPU_PRIORITY_VAL:STRING=""
  OBS_CMAKE_VERSION:STRING="3.0.0"
  RESTREAM_CLIENTID:STRING=""
  RESTREAM_HASH:STRING=""
  TWITCH_CLIENTID:STRING=""
  TWITCH_HASH:STRING=""
  VIRTUALCAM_GUID:STRING="A3FCE0F5-3493-419F-958A-ABA1250EC20B"
  YOUTUBE_CLIENTID:STRING=""
  YOUTUBE_CLIENTID_HASH:STRING=""
  YOUTUBE_SECRET:STRING=""
  YOUTUBE_SECRET_HASH:STRING=""

CMake Error at cmake/common/versionconfig.cmake:19 (message):
  Could not fetch OBS version tag from git.

  fatal: No names found, cannot describe anything.

Call Stack (most recent call first):
  cmake/common/bootstrap.cmake:37 (include)
  CMakeLists.txt:4 (include)


-- Configuring incomplete, errors occurred!
```

PatTheMav commented:
Our documentation clearly states to check out the project using `git clone --recursive https://github.com/obsproject/obs-studio.git`, which will fetch all tags and recursively check out all submodules.

Does the issue that this PR tries to solve occur despite cloning the repo in the specified way?

Aleksbgbg commented:
For the first commit - yes, git may fail to give a description for a repo for various reasons even when the repo is cloned correctly, most notably when the repository is marked as unsafe by git which can occur for some users (sharing a file system with another operating system, user, or virtualization manager (Docker, VirtualBox, etc)).

For the second commit, we to give a better error message in the situation where tags were not fetched. This could also happen when tags were accidentally deleted even after cloning correctly, for example, but I can remove it if it is not desired.

PatTheMav commented:
> For the first commit - yes, git may fail to give a description for a repo for various reasons even when the repo is cloned correctly, most notably when the repository is marked as unsafe by git which can occur for some users (sharing a file system with another operating system, user, or virtualization manager (Docker, VirtualBox, etc)).

What does "marked as unsafe by git" mean? What's the associated error message?

> For the second commit, we aim to give a better error message in the situation where tags were not fetched. This could also happen when tags were accidentally deleted even after cloning correctly, for example, but I can remove it if it is not desired.

If a user deletes the tags, then I expect them to be able to fetch them again. If a user decides to experiment with git commands, then it's their responsibility to understand what a command does and take the responsibility to fix it.

My main issue is that changes like these make it the OBS' maintainers responsibility to cover for developer mistakes and errors, making it a maintenance burden (because it becomes part of the codebase and as such something we need to maintain and keep in working condition) even though it's entirely the user's responsibility to understand their local environment and fix their own mistakes.

Aleksbgbg commented:
> What does "marked as unsafe by git" mean? What's the associated error message?
```
  fatal: detected dubious ownership in repository at '/src/obs-studio'

  To add an exception for this directory, call:
        git config --global --add safe.directory /src/obs-studio
```

Explained in detail in https://git.kernel.org/pub/scm/git/git.git/commit/?id=8959555cee7ec045958f9b6dd62e541affb7e7d9. In summary, you can engineer git config that causes git to execute arbitrary code when the user first opens their shell, so git disables itself on directories you do not own. until you mark them as safe.

If this happens to someone developing OBS in a directory not owned by them, they get a cryptic build error instead of the git error message above, which is fixed by the first commit. The before and after is shown in the opening comment.

> My main issue is that changes like these make it the OBS' maintainers responsibility to cover for developer mistakes and errors, making it a maintenance burden (because it becomes part of the codebase and as such something we need to maintain and keep in working condition) even though it's entirely the user's responsibility to understand their local environment and fix their own mistakes.

Sure, that's fair. Although, with the frequency at which people ask questions with an incorrect clone, it's not entirely clear whether the maintenance burden of a clear error message is actually cheaper than answering all of the questions (a classic case of "not my fault but still my problem"). Call me back if you do want this in the future 😄 .

So, I'll leave the decision up to you - I've left only the fix for git error (such as unsafe directories). If you think this is useful, feel free to merge it. If you do not want the maintenance burden, feel free to close the pull request.

PatTheMav commented:
I think the current scope of the PR (printing the error message by `git`) should suffice because it converts an opaque error ("something failed") into something actionable ("somewhere git failed with this error") as it doesn't require knowledge of how/where the build system invokes git - it's sufficient to know that it does and that the repository needs to be fixed up.

RytoEX reviewed (changes requested):
Commit message nit:
* colour-enabled -> color-enabled

PR title nit:
Use the commit message subject, including prefix.

Otherwise seems fine.

RytoEX reviewed (approved):
Seems fine.
